### DISCUSSION TITLE
Is running sql_query() with a "column IN array" syntax supported?

### DISCUSSION THREAD

**Discussion by randall-coding** [diesel-rs/diesel#3652] [Q&A]
This seems like a standard sql feature so I'm assuming there is some way to do this.  But so far I cannot figure it out.

My code: 

```
let query = diesel::sql_query("SELECT * FROM bids t1
    LEFT JOIN ab_matches t2 ON t2.bid_id = t1.id
    WHERE t2.bid_id IS NULL
    AND t1.color IN ? OR ? IS NULL
    ");
  let bids: Result<Vec<Bid>, diesel::result::Error> = query
      .bind::<Vec<Text>, _>(&color)
      .bind::<Vec<Text>, _>(&color)
      .get_results(conn);
```

Using `Vec<Text>` instead of the usual `<Text>` on bind() yields a compile error:

> error[E0277]: the trait bound Vec\<diesel::sql_types::Text\>: QueryId is not satisfied
> --> src/db/[bids.rs:140](http://bids.rs:140/):20
> 
> 140 | .get_results(conn);
> ----------- ^^^^ the trait QueryId is not implemented for Vec\<diesel::sql_types::Text\>
>

**Comment by weiznich** [comment]
This code snippet contain quite a few issues, it's hard to decide which one to start with. It's close to the point where it's even unclear what you are trying to do at all as your question is missing a lot of context (database system, diesel version, why you want not use the dsl, …). Therefore I will just leave a list of issues here:

* [`bind`](https://docs.diesel.rs/2.1.x/diesel/query_builder/struct.SqlQuery.html#method.bind) accepts a sql type as first parameter, `Vec<Text>` is not an sql type.
* Your bind syntax indicates that you use either mysql or sqlite. Both do not support binding "arrays"
* `IN` always expects a bind parameter per entry in the check list.  Also you miss the parenthesis. The correct syntax is `IN (?, ?, …, ?)`. 
* Checking whether or not a non-nullable bind is null is doing duplicated work (the `? IS NULL` part)

**Reply by randall-coding** [reply]
I'm using Sqlite and Diesel 2.0.4.

I was looking for functionality that would replace the `?` with the proper syntax for a list, the `IN (value, value, value)` that you alluded to.

If that isn't what sql_query was designed for, my question is what is the proper way to handle a dynamic Vec of values as a filter parameter in your dsl?  

My goal is to use diesel to make queries safely and more conveniently than generating sql purely from scratch.  Would you recommend this library for that or another one?

**Reply by weiznich** [reply]
> I was looking for functionality that would replace the ? with the proper syntax for a list, the IN (value, value, value) that you alluded to.

To write that out explicitly: `sql_query` is designed to accept a sql query string and pass it as provided to the database. It does **not** perform any changes to the provided string.
As for the proper syntax: I've already written that above: You need a single bind per entry in your vector. (Honestly constructing `IN` expressions is something query builders excel at, while it's a pain to do that with sql strings).

> If that isn't what sql_query was designed for, my question is what is the proper way to handle a dynamic Vec of values as a filter parameter in your dsl?

Maybe don't use `sql_query` then, but the built-in DSL. If you search the API documentation for [`IN`](https://docs.diesel.rs/2.1.x/diesel/index.html?search=in) you find the correct methods easily. (That works as well for most other SQL constructs)

**Reply by randall-coding** [reply]
Thanks.  Yes I think searching API docs would be preferable.  Certainly better than getting support from less than sociable open source devs, who make you feel dumb for daring to ask a question when you get stuck.

**Reply by weiznich** [reply]
Well its as simple as follows: There are more than you as a diesel user that wants support for something. In addition I provide this support in my free time for free. So if you now ask a question thats basically just a dump of incomplete code with an error message you basically ask me to figure it out on my own whatever you might want to do there. That takes a lot of time away from support for other peoples or working on diesel improvements.
 I normally answer questions with a similar amount of effort that the question asker spend on writing their question. So if you feel that an answer is unfriendly, short or incomplete (=only contains pointers where to look) that might be an issue with your question.

If your really want that level of support where you can just dump such questions on me I suggest that we talk about a support contract.
